esphome/components/shelly_dimmer/stm32flash.cpp

```diff
@@ -0,0 +1,1042 @@
+    return STM32_ERR_UNKNOWN;
+
+  return stm32_check_ack_timeout(stm32_get_ack_timeout(stm, STM32_WUNPROT_TIMEOUT),
+                                 []() { ESP_LOGD(TAG, "Error: Failed to WRITE UNPROTECT"); });
+}
+
+stm32_err_t stm32_wprot_memory(const stm32_t *stm) {
+  if (stm->cmd->wp == STM32_CMD_ERR) {
+    ESP_LOGD(TAG, "Error: WRITE PROTECT command not implemented in bootloader.");
+    return STM32_ERR_NO_CMD;
+  }
+
+  if (stm32_send_command(stm, stm->cmd->wp) != STM32_ERR_OK)
+    return STM32_ERR_UNKNOWN;
+
+  return stm32_check_ack_timeout(stm32_get_ack_timeout(stm, STM32_WPROT_TIMEOUT),
+                                 []() { ESP_LOGD(TAG, "Error: Failed to WRITE PROTECT"); });
+}
+
+stm32_err_t stm32_runprot_memory(const stm32_t *stm) {
+  if (stm->cmd->ur == STM32_CMD_ERR) {
+    ESP_LOGD(TAG, "Error: READOUT UNPROTECT command not implemented in bootloader.");
+    return STM32_ERR_NO_CMD;
+  }
+
+  if (stm32_send_command(stm, stm->cmd->ur) != STM32_ERR_OK)
+    return STM32_ERR_UNKNOWN;
+
+  return stm32_check_ack_timeout(stm32_get_ack_timeout(stm, STM32_MASSERASE_TIMEOUT),
+                                 []() { ESP_LOGD(TAG, "Error: Failed to READOUT UNPROTECT"); });
+}
+
+stm32_err_t stm32_readprot_memory(const stm32_t *stm) {
+  if (stm->cmd->rp == STM32_CMD_ERR) {
+    ESP_LOGD(TAG, "Error: READOUT PROTECT command not implemented in bootloader.");
+    return STM32_ERR_NO_CMD;
+  }
+
+  if (stm32_send_command(stm, stm->cmd->rp) != STM32_ERR_OK)
+    return STM32_ERR_UNKNOWN;
+
+  return stm32_check_ack_timeout(stm32_get_ack_timeout(stm, STM32_RPROT_TIMEOUT),
+                                 []() { ESP_LOGD(TAG, "Error: Failed to READOUT PROTECT"); });
+}
+
+stm32_err_t stm32_erase_memory(const stm32_t *stm, uint32_t spage, uint32_t pages) {
+  if (!pages || spage > STM32_MAX_PAGES || ((pages != STM32_MASS_ERASE) && ((spage + pages) > STM32_MAX_PAGES)))
+    return STM32_ERR_OK;
+
+  if (stm->cmd->er == STM32_CMD_ERR) {
+    ESP_LOGD(TAG, "Error: ERASE command not implemented in bootloader.");
+    return STM32_ERR_NO_CMD;
+  }
+
+  if (pages == STM32_MASS_ERASE) {
+    /*
+     * Not all chips support mass erase.
+     * Mass erase can be obtained executing a "readout protect"
+     * followed by "readout un-protect". This method is not
+     * suggested because can hang the target if a debug SWD/JTAG
+     * is connected. When the target enters in "readout
+     * protection" mode it will consider the debug connection as
+     * a tentative of intrusion and will hang.
+     * Erasing the flash page-by-page is the safer way to go.
+     */
+    if (!(stm->dev->flags & F_NO_ME))
+      return stm32_mass_erase(stm);
+
+    pages = flash_addr_to_page_ceil(stm, stm->dev->fl_end);
+  }
+
+  /*
+   * Some device, like STM32L152, cannot erase more than 512 pages in
+   * one command. Split the call.
+   */
+  static constexpr uint32_t MAX_PAGE_SIZE = 512;
+  while (pages) {
+    const auto n = std::max(pages, MAX_PAGE_SIZE);
```

Comment:
```suggestion
    const auto n = std::min(pages, MAX_PAGE_SIZE);
```

https://github.com/wichers/stm32flash/blob/master/stm32flash.cpp#L894